src/Symfony/Bundle/SecurityBundle/Command/DebugFirewallCommand.php

```diff
@@ -0,0 +1,262 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\SecurityBundle\Command;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Bundle\SecurityBundle\Security\FirewallContext;
+use Symfony\Bundle\SecurityBundle\Security\LazyFirewallContext;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * @author Timo Bakx <timobakx@gmail.com>
+ */
+final class DebugFirewallCommand extends Command
+{
+    protected static $defaultName = 'debug:firewall';
+
+    /**
+     * @var string[]
+     */
+    private $firewallNames;
+
+    /**
+     * @var ContainerInterface
+     */
+    private $contexts;
+
+    /**
+     * @var ContainerInterface
+     */
+    private $eventDispatchers;
+
+    /**
+     * @var
+     */
+    private $authenticators;
+
+    public function __construct(
+        array $firewallNames,
+        ContainerInterface $contexts,
+        ContainerInterface $eventDispatchers,
+        $authenticators
+    ) {
+        parent::__construct();
+
+        $this->firewallNames = $firewallNames;
+        $this->contexts = $contexts;
+        $this->eventDispatchers = $eventDispatchers;
+        $this->authenticators = $authenticators;
+    }
+
+    protected function configure(): void
+    {
+        $this
+            ->setDefinition([
+                new InputArgument('name', InputArgument::OPTIONAL, 'A firewall name (main)'),
+            ]);
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $io = new SymfonyStyle($input, $output);
+
+        $name = $input->getArgument('name');
+
+        if (null === $name) {
+            $this->displayFirewallList($io);
+
+            return 0;
+        }
+
+        $serviceId = sprintf('security.firewall.map.context.%s', $name);
+
+        if (!$this->contexts->has($serviceId)) {
+            $io->error(sprintf(
+                'Firewall %s was not found. Available firewalls are: %s',
+                $name,
+                implode(', ', $this->firewallNames)
+            ));
+
+            return 1;
+        }
+
+        /** @var FirewallContext $context */
+        $context = $this->contexts->get($serviceId);
+
+        $io->title(sprintf('Firewall "%s"', $name));
+
+        $this->displayFirewallSummary($name, $context, $io);
+
+        $this->displaySwitchUser($context, $io);
+
+        $this->displayEventListeners($name, $context, $io);
```

Comment:
Maybe we need an option to show the listeners and otherwise hide them (like service arguments in DI)? They take up quite a lot of lines

Reply:
Good idea! I'll add it to the TODO's.

Reply:
This is now optional and will be displayed by using either `-l` or `--include-listeners`.